File: sdk/python/polos/core/context.py
"""Context classes for workflow and agent execution."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel


class WorkflowContext:
    """Context available to all workflow functions.

    Provides execution-related information that workflows can use for
    checkpointing, triggering sub-workflows, and other operations.

    Includes state managers for workflow-scoped, session-scoped, and user-scoped state.
    Also provides a Step helper for durable execution.
    """

    def __init__(
        self,
        workflow_id: str,
        execution_id: str,
        deployment_id: str,
        session_id: str,
        user_id: str | None = None,
        parent_execution_id: str | None = None,
        root_execution_id: str | None = None,
        root_workflow_id: str | None = None,
        retry_count: int = 0,
        created_at: datetime | None = None,
        workflow_type: str | None = "workflow",
        otel_traceparent: str | None = None,
        otel_span_id: str | None = None,
        state_schema: type[BaseModel] | None = None,
        initial_state: dict[str, Any] | None = None,
    ):
        self.workflow_id = workflow_id
        self.execution_id = execution_id
        self.deployment_id = deployment_id
        self.parent_execution_id = parent_execution_id
        self.root_execution_id = root_execution_id or execution_id
        self.root_workflow_id = root_workflow_id or workflow_id
        self.retry_count = retry_count
        self.created_at = created_at
        self.session_id = session_id
        self.user_id = user_id
        self.workflow_type = workflow_type
        self.otel_traceparent = otel_traceparent
        self.otel_span_id = otel_span_id

        # Initialize typed workflow state if state_schema is provided
        if state_schema:
            if initial_state:
                self.state = state_schema.model_validate(initial_state)
            else:
                self.state = state_schema()  # Use defaults
        else:
            self.state = None

        # Initialize step helper for durable execution
        from .step import Step

        self.step = Step(self)

    def to_dict(self) -> dict[str, Any]:
        """Convert context to dictionary for execution context cache."""
        return {
            "workflow_id": self.workflow_id,
            "execution_id": self.execution_id,
            "deployment_id": self.deployment_id,
            "parent_execution_id": self.parent_execution_id,
            "root_execution_id": self.root_execution_id,
            "retry_count": self.retry_count,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "session_id": self.session_id,
            "user_id": self.user_id,
        }


class AgentContext(WorkflowContext):
    """Context for agents - extends WorkflowContext with agent information."""

    def __init__(
        self,
        agent_id: str,
        execution_id: str,
        deployment_id: str,
        parent_execution_id: str | None = None,
        root_execution_id: str | None = None,
        root_workflow_id: str | None = None,
        retry_count: int = 0,
        model: str = "gpt-4",
        provider: str = "openai",
        system_prompt: str | None = None,
        tools: list[Any] | None = None,
        temperature: float | None = None,
        max_tokens: int | None = None,
        session_id: str | None = None,
        conversation_id: str | None = None,
        user_id: str | None = None,
        created_at: datetime | None = None,
        otel_traceparent: str | None = None,
        otel_span_id: str | None = None,
        state_schema: type[BaseModel] | None = None,
        initial_state: dict[str, Any] | None = None,
    ):
        # Call parent with required parameters
        super().__init__(
            workflow_id=agent_id,
            execution_id=execution_id,
            deployment_id=deployment_id,
            session_id=session_id,
            user_id=user_id,
            parent_execution_id=parent_execution_id,
            root_execution_id=root_execution_id,
            root_workflow_id=root_workflow_id,
            retry_count=retry_count,
            created_at=created_at,
            workflow_type="agent",
            otel_traceparent=otel_traceparent,
            otel_span_id=otel_span_id,
            state_schema=state_schema,
            initial_state=initial_state,
        )
        self.agent_id = agent_id
        self.model = model
        self.provider = provider
        self.system_prompt = system_prompt
        self.tools = tools or []
        self.temperature = temperature
        self.max_tokens = max_tokens
        self.conversation_id = conversation_id

    def to_dict(self) -> dict[str, Any]:
        """Convert context to dictionary, including agent-specific fields."""
        base_dict = super().to_dict()
        base_dict.update(
            {
                "model": self.model,
                "provider": self.provider,
                "system_prompt": self.system_prompt,
                "tools": self.tools,
                "temperature": self.temperature,
                "max_tokens": self.max_tokens,
            }
        )
        return base_dict

File: sdk/typescript/src/core/context.ts
/**
 * WorkflowContext implementation.
 *
 * Provides the execution context for workflows including state access,
 * step execution, and metadata about the current execution.
 */

import type { StepHelper } from './step.js';

/**
 * Context available during workflow execution.
 */
export interface WorkflowContext<TState = unknown> {
  /** Workflow identifier */
  readonly workflowId: string;
  /** Unique execution identifier */
  readonly executionId: string;
  /** Deployment identifier */
  readonly deploymentId: string;
  /** Session identifier (if provided) */
  readonly sessionId?: string | undefined;
  /** User identifier (if provided) */
  readonly userId?: string | undefined;
  /** Parent execution identifier (if this is a sub-workflow) */
  readonly parentExecutionId?: string | undefined;
  /** Root execution identifier (top-level execution in the hierarchy) */
  readonly rootExecutionId: string;
  /** Root workflow identifier (top-level workflow in the hierarchy) */
  readonly rootWorkflowId: string;
  /** Number of times this execution has been retried */
  readonly retryCount: number;
  /** When the execution was created (ISO 8601 string) */
  readonly createdAt?: string | undefined;
  /** Workflow type: "workflow", "agent", or "tool" */
  readonly workflowType?: string | undefined;
  /** OpenTelemetry trace parent */
  readonly otelTraceparent?: string | undefined;
  /** OpenTelemetry span ID */
  readonly otelSpanId?: string | undefined;
  /** Workflow state (persisted across executions) */
  state: TState;
  /** Step helper for durable operations */
  readonly step: StepHelper;
}

/**
 * Context for agent execution — extends WorkflowContext with agent-specific fields.
 * Matches Python AgentContext(WorkflowContext).
 */
export interface AgentContext<TState = unknown> extends WorkflowContext<TState> {
  /** Agent identifier (same as workflowId) */
  readonly agentId: string;
  /** LLM model identifier */
  readonly model: string;
  /** LLM provider identifier */
  readonly provider: string;
  /** System prompt */
  readonly systemPrompt?: string | undefined;
  /** Tools available to the agent */
  readonly tools: unknown[];
  /** Temperature for LLM generation */
  readonly temperature?: number | undefined;
  /** Maximum output tokens */
  readonly maxTokens?: number | undefined;
  /** Conversation ID (mutable — set by agent handler, matching Python) */
  conversationId?: string | undefined;
}

/**
 * Internal context data used during workflow execution.
 */
export interface WorkflowContextData<TState = unknown> {
  workflowId: string;
  executionId: string;
  deploymentId: string;
  sessionId?: string | undefined;
  userId?: string | undefined;
  parentExecutionId?: string | undefined;
  rootExecutionId: string;
  rootWorkflowId: string;
  retryCount: number;
  createdAt?: string | undefined;
  workflowType?: string | undefined;
  otelTraceparent?: string | undefined;
  otelSpanId?: string | undefined;
  state: TState;
}

/**
 * Options for creating a workflow context.
 */
export interface CreateContextOptions<TState> {
  /** Workflow identifier */
  workflowId: string;
  /** Execution identifier (generated if not provided) */
  executionId?: string | undefined;
  /** Deployment identifier */
  deploymentId: string;
  /** Session identifier */
  sessionId?: string | undefined;
  /** User identifier */
  userId?: string | undefined;
  /** Parent execution identifier */
  parentExecutionId?: string | undefined;
  /** Root execution identifier (defaults to executionId) */
  rootExecutionId?: string | undefined;
  /** Root workflow identifier (defaults to workflowId) */
  rootWorkflowId?: string | undefined;
  /** Retry count */
  retryCount?: number | undefined;
  /** When the execution was created */
  createdAt?: string | undefined;
  /** Workflow type: "workflow", "agent", or "tool" */
  workflowType?: string | undefined;
  /** OpenTelemetry trace parent */
  otelTraceparent?: string | undefined;
  /** OpenTelemetry span ID */
  otelSpanId?: string | undefined;
  /** Initial state */
  initialState: TState;
  /** Step helper instance */
  stepHelper: StepHelper;
}

/**
 * Create a workflow context.
 */
export function createWorkflowContext<TState>(
  options: CreateContextOptions<TState>
): WorkflowContext<TState> {
  const executionId = options.executionId ?? crypto.randomUUID();

  const context: WorkflowContext<TState> = {
    workflowId: options.workflowId,
    executionId,
    deploymentId: options.deploymentId,
    sessionId: options.sessionId,
    userId: options.userId,
    parentExecutionId: options.parentExecutionId,
    rootExecutionId: options.rootExecutionId ?? executionId,
    rootWorkflowId: options.rootWorkflowId ?? options.workflowId,
    retryCount: options.retryCount ?? 0,
    createdAt: options.createdAt,
    workflowType: options.workflowType ?? 'workflow',
    otelTraceparent: options.otelTraceparent,
    otelSpanId: options.otelSpanId,
    state: options.initialState,
    step: options.stepHelper,
  };

  return context;
}

/**
 * Extract context data for serialization.
 */
export function extractContextData<TState>(
  ctx: WorkflowContext<TState>
): WorkflowContextData<TState> {
  return {
    workflowId: ctx.workflowId,
    executionId: ctx.executionId,
    deploymentId: ctx.deploymentId,
    sessionId: ctx.sessionId,
    userId: ctx.userId,
    parentExecutionId: ctx.parentExecutionId,
    rootExecutionId: ctx.rootExecutionId,
    rootWorkflowId: ctx.rootWorkflowId,
    retryCount: ctx.retryCount,
    createdAt: ctx.createdAt,
    workflowType: ctx.workflowType,
    otelTraceparent: ctx.otelTraceparent,
    otelSpanId: ctx.otelSpanId,
    state: ctx.state,
  };
}
